hazelcast/src/main/java/com/hazelcast/transaction/impl/TransactionLog.java

```diff
@@ -38,53 +37,40 @@
  * the TransactionLog will contains 4 {@link TransactionLogRecord} instances.
  *
  * planned optimization:
- * Most transaction will be small, but an linkedlist + hashmap is created. Instead use an array and do a
+ * Most transaction will be small, but an linkedhashmap is created. Instead use an array and do a
  * linear search in that array. When there are too many items added, then enable the hashmap.
  */
 public class TransactionLog {
 
-    private final List<TransactionLogRecord> recordList = new LinkedList<TransactionLogRecord>();
-    private final Map<Object, TransactionLogRecord> recordMap = new HashMap<Object, TransactionLogRecord>();
+    private final Map<Object, TransactionLogRecord> recordMap = new LinkedHashMap<>();
```

Comment:
Why not to use simple `HashMap` here which is more memory efficient? Every key is independent and we don't have to support any order across the keys.

Reply:
I wasn't sure about that. As you can see, previous implementation was based on a `LinkedList` + `HashMap`, and in methods `commit()` and `prepare()`, the records were iterated in the order in which they were added to the transaction log. I wanted to preserve this behavior in case something relied on this. Especially that in `rollback()` method, in previous implementation, the iteration was purposely done in *reverse* order, so that was a hint that the order matters. See this discussion: https://github.com/hazelcast/hazelcast/pull/15111#discussion_r288922671, where I asked about whether the reverse order is important.

If you confirm that the order of iteration of records does not matter, I'll update this PR.

Reply:
@gurbuzali, can you also confirm that the order of iteration across the keys doesn't matter? I personally don't see why would we need the order because the key/value pairs are completely independent.

Reply:
Yes, the order should not be of importance here. They are independent

Reply:
I must admit I haven't looked at the code too deep but seeing the code iterate in order gave me the gut feeling that order does matter, as @gdela mentioned. 
But since I see the loops invoke operations asynchronously and since you two mention order does not matter, I'm going to agree with you that order does not matter.

Reply:
I've changed the transaction log to use plain `HashMap`, see 416bca273a13eae8aaf8f17ccfcf17c14797a544. Looks like nothing bad happened, our application that uses Hazelcast with this patch still works OK.

BTW, my colleague has just found out, that either we are doing something wrong, or Hazelcast transactions are not guaranteed to be atomic. See #15542 where we show a reproduction of this non-atomicity (this is on clean version of Hazelcast, without this PR applied).